### PR TITLE
Resent50 Notebook

### PR DESCRIPTION
**Pull request by farazkh80**

### CODE REVIEW

**Comment by parthchadha** [comment]
1. Lets remove TP prefix from all modules.
2. There is no training with tripy so you can remove the `if training` code from BatchNorm.

**Comment by parthchadha** [comment]
```
def get_tp_pred(model, image, processor):
    processed_image = processor(image, return_tensors="np")['pixel_values']
    tp_image = tp.Tensor(processed_image.squeeze(), dtype=tp.float32, device=tp.device("gpu"))
    tp_image = tp.unsqueeze(tp_image, dim=0)
    return tp.argmax(model(tp_image), -1).tolist()[0]
```

^ uses Tripy outside of the compiled model inference. Even though this is functionally correct, running Tripy operations in eager mode can be expensive and I would rather use torch for pre and post processing operations.

**Comment by farazkh80** [comment]
Thanks for the comments @parthchadha.

I removed `TP` prefix and extra training mode code from `BatchNorm`. I also did the pre/post processing steps with `torch` to avoid `tripy` eager mode.

File: tripy/examples/resent50/tripy-res50.ipynb [hunk 1]
```diff
@@ -0,0 +1,599 @@
...
+    "        x = self.weight * x + self.bias\n",
+    "\n",
+    "        return x\n",
+    "\n",
+    "class Identity(tp.Module):\n",
+    "    def __init__(self):\n",
+    "        super(Identity, self).__init__()\n",
+    "\n",
+    "    def __call__(self, x):\n",
+    "        return x\n",
```
Comment (pranavm-nvidia) [inline tripy/examples/resent50/tripy-res50.ipynb:139]: Do we need an identity module?

File: tripy/examples/resent50/tripy-res50.ipynb [hunk 2]
```diff
@@ -0,0 +1,599 @@
...
+   "source": [
+    "class MaxPool2d(tp.Module):\n",
+    "    def __init__(self, kernel_dims, stride=None, padding=((0, 0), (0, 0))):\n",
+    "        super(MaxPool2d, self).__init__()\n",
+    "        self.kernel_dims = kernel_dims\n",
+    "        self.stride = stride if stride is not None else kernel_dims\n",
+    "        self.padding = padding\n",
+    "\n",
+    "    def __call__(self, x):\n",
+    "        return tp.maxpool(x, kernel_dims=self.kernel_dims, stride=self.stride, padding=self.padding)\n",
```
Comment (pranavm-nvidia) [inline tripy/examples/resent50/tripy-res50.ipynb:268]: Can we use `tp.maxpool` directly instead of wrapping it in a module?

File: tripy/examples/resent50/tripy-res50.ipynb [hunk 3]
```diff
@@ -0,0 +1,599 @@
...
+    "        H_out, W_out = self.output_size\n",
+    "\n",
+    "        # Calculate stride and kernel size\n",
+    "        stride_h = H_in // H_out\n",
+    "        stride_w = W_in // W_out\n",
+    "\n",
+    "        kernel_size_h = H_in - (H_out - 1) * stride_h\n",
+    "        kernel_size_w = W_in - (W_out - 1) * stride_w\n",
+    "\n",
+    "        return tp.avgpool(x, kernel_dims=(int(kernel_size_h), int(kernel_size_w)), stride=(int(stride_h), int(stride_w)))\n",
```
Comment (pranavm-nvidia) [inline tripy/examples/resent50/tripy-res50.ipynb:312]: Could we just make this a helper function instead of a module?

File: tripy/examples/resent50/tripy-res50.ipynb [hunk 4]
```diff
@@ -0,0 +1,599 @@
...
+   "outputs": [],
+   "source": [
+    "class Flatten(tp.Module):\n",
+    "    def __init__(self, start_dim=1, end_dim=-1):\n",
+    "        super(Flatten, self).__init__()\n",
+    "        self.start_dim = start_dim\n",
+    "        self.end_dim = end_dim\n",
+    "\n",
+    "    def __call__(self, x):\n",
+    "        return tp.flatten(x, start_dim=self.start_dim, end_dim=self.end_dim)\n",
```
Comment (pranavm-nvidia) [inline tripy/examples/resent50/tripy-res50.ipynb:356]: Let's just use `tp.flatten` instead of creating a module?

File: tripy/examples/resent50/tripy-res50.ipynb (hunk 1 above)
Comment (farazkh80) [reply tripy/examples/resent50/tripy-res50.ipynb:139]: removed

File: tripy/examples/resent50/tripy-res50.ipynb (hunk 2 above)
Comment (farazkh80) [reply tripy/examples/resent50/tripy-res50.ipynb:268]: done, removed module

File: tripy/examples/resent50/tripy-res50.ipynb (hunk 3 above)
Comment (farazkh80) [reply tripy/examples/resent50/tripy-res50.ipynb:312]: done

File: tripy/examples/resent50/tripy-res50.ipynb (hunk 4 above)
Comment (farazkh80) [reply tripy/examples/resent50/tripy-res50.ipynb:356]: done

**Comment by farazkh80** [comment]
Hey @pranavm-nvidia @parthchadha 

- Addressed the current suggested changes in the PR, and removed unnecessary modules from the notebook. 
- Added `BatchNorm` as a module similar to `LayerNorm` and `GroupNorm` under `module`

**Comment by farazkh80** [comment]
closing since I have a feature branch open now
